dev/tools/localization/gen_l10n.dart

```diff
@@ -302,184 +314,230 @@ bool _isValidGetterAndMethodName(String name) {
+      );
   }
 
-  final RegExp pluralValueRE = RegExp(r'^\s*\{[\w\s,]*,\s*plural\s*,');
+  /// Sets the reference [File] for [templateArbFile].
+  void setTemplateArbFile(String templateArbFileName) {
+    if (templateArbFileName == null)
+      throw L10nException('templateArbFileName argument cannot be null');
+    if (l10nDirectory == null)
+      throw L10nException('l10nDirectory cannot be null when setting template arb file');
+
+    templateArbFile = _fs.file(path.join(l10nDirectory.path, templateArbFileName));
+    final String templateArbFileStatModeString = templateArbFile.statSync().modeString();
+    if (templateArbFileStatModeString[0] == '-' && templateArbFileStatModeString[3] == '-')
+      throw FileSystemException(
+        "The 'template-arb-file', $templateArbFile, is not readable.\n"
+        'Please ensure that the user has read permissions.'
+      );
+  }
+
+  /// Sets the reference [File] for the localizations delegate [outputFile].
+  void setOutputFile(String outputFileString) {
+    if (outputFileString == null)
+      throw L10nException('outputFileString argument cannot be null');
+    outputFile = _fs.file(path.join(l10nDirectory.path, outputFileString));
+  }
 
-  for (String key in bundle.keys.toList()..sort()) {
-    if (key.startsWith('@'))
-      continue;
-    if (!_isValidGetterAndMethodName(key))
-      exitWithError(
-        'Invalid key format: $key \n It has to be in camel case, cannot start '
-        'with a number, and cannot contain non-alphanumeric characters.'
+  /// Sets the [className] for the localizations and localizations delegate
+  /// classes.
+  void setClassName(String classNameString) {
+    if (classNameString == null)
+      throw L10nException('classNameString argument cannot be null');
+    if (!_isValidClassName(classNameString))
+      throw L10nException(
+        "The 'output-class', $classNameString, is not a valid Dart class name.\n"
       );
-    if (pluralValueRE.hasMatch(bundle[key]))
-      classMethods.add(genPluralMethod(bundle, key));
-    else
-      classMethods.add(genSimpleMethod(bundle, key));
+    className = classNameString;
   }
 
-  outputFile.writeAsStringSync(
-    defaultFileTemplate
-      .replaceAll('@className', stringsClassName)
-      .replaceAll('@classMethods', classMethods.join('\n'))
-      .replaceAll('@importFile', _importFilePath(arbPathString, outputFileString))
-      .replaceAll('@supportedLocales', genSupportedLocaleProperty(supportedLocales))
-      .replaceAll('@supportedLanguageCodes', supportedLanguageCodes.toList().join(', '))
-  );
-
-  final ProcessResult pubGetResult = await Process.run('flutter', <String>['pub', 'get']);
-  if (pubGetResult.exitCode != 0) {
-    stderr.write(pubGetResult.stderr);
-    exit(1);
+  /// Scans [l10nDirectory] for arb files and parses them for language and locale
+  /// information.
+  void parseArbFiles() {
+    final List<FileSystemEntity> fileSystemEntityList = l10nDirectory
+      .listSync()
```

Comment:
Add a `.whereType<File>` and you can remove the `entity is File` check nesting below